src/core/dbus-manager.c

```diff
@@ -1409,16 +1409,37 @@ static int dump_impl(
 
         _cleanup_free_ char *dump = NULL;
         Manager *m = ASSERT_PTR(userdata);
+        bool rate_limited= false;
         int r;
 
         assert(message);
 
-        /* Anyone can call this method */
-
+        /* 'status' access is the bare minimum always needed for this */
         r = mac_selinux_access_check(message, "status", error);
         if (r < 0)
                 return r;
 
+        /* Reload/reexec are also slow but non-destructive, use the same access check for dumps which, given
+         * the large amount of data to fetch, can stall for quite some time. If the caller is not privileged
+         * use rate limiting. */
+        r = mac_selinux_access_check(message, "reload", error);
+        if (r < 0)
+                rate_limited = true;
+
+        r = bus_verify_reload_daemon_async(m, message, error);
+        if (r < 0)
+                rate_limited = true;
+        if (r == 0)
+                return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
+
+        /* Check the rate limit after the authorization succeeds, to avoid denial-of-service issues. */
+        if (rate_limited && !ratelimit_below(&m->dump_ratelimit)) {
+                log_warning("Dump request rejected due to rate limit.");
+                return sd_bus_error_setf(error,
```

Comment:
Instead of refusing the request, how about to cache the previous dump and provide it?
This is just a debugging interface, returning slightly older info should not cause any issue.

Reply:
I'm not sure, when debugging giving out stale information silently seems way worse to me, than a clear error message

Reply:
also it's a lot of memory, and allowing an unprivileged user to keep the memory footprint permanently inflated doesn't seem good either

Reply:
I think just introducing a timer event source for the cached result is enough. Then, a certain time passed, we can free the cached result. If a privileged request comes, then we can always replace the existing cached result with new one. So, there exists only one cached result per manager at most.

Currently, we already require an amount of memory to store the dump result to bus message at least tentatively. So, the memory footprint should not change. Why do you think we should not keep the memory for a certain time period?

Reply:
> I'm not sure, when debugging giving out stale information silently seems way worse to me, than a clear error message

If the way I suggest is accepted, then we should add the timestamp that the dump result is generated.
That should be enough, as this is a debugging interface.

Reply:
> Currently, we already require an amount of memory to store the dump result to bus message at least tentatively. So, the memory footprint should not change. Why do you think we should not keep the memory for a certain time period?

Nah, that only happens in-flight, for an extremely brief period of time. It is wildly different from permanently doubling the memory footprint. I do not see any reason to introduce such complication.

Reply:
That's debugging interface, and usually debugging something requires much much memory.

> It is wildly different from permanently doubling the memory footprint.

Not permanently, for a certain time period, say 10sec or so.

Reply:
Anyway, that's not a strong opinion, and just a suggestion. I prefer to cache the result, and make Debug() always work even for unprivileged users even the info is slightly outdated, but if others agree with the current form, then that's OK.

Reply:
> Not permanently, for a certain time period, say 10sec or so.

Yes but it can be repeated forever if there's no rate limiting - this is a DOS vector given it doesn't take privileges

Reply:
Yes of course, but the amount does not increase.

Reply:
it's double what it would be with the current approach, and on small servers or embedded systems that can be enough to trigger pressure and OOMs

Reply:
That's theoretically true, but is it realistic?
On my laptop, `systemd-analyze dump` outputs only 1.0MB. So, on small servers or embedded machines, I guess that's at most ~< 10MB so, in a worst case. If caching dump result causes memory pressure, then the system can easily trigger OOM by other reasons. I'd like to say that if Dump() causes OOM, then the design of the system is completely borked.

Reply:
An extra 10M is doubling the usual memory footprint. And yes, this happens regularly. Allowing unprivileged users to do this is not acceptable. Local privileged users can simply use 'sudo' and get the existing behaviour with no differences.

Reply:
Hmm, I think the current implementation makes more sense. It sounds spurious to me that the interface could return different info depending on the privilege.

Reply:
dump is a debugging tool. I don't think people would be too happy if the data it reports would be out-of-date, because cached.